### ISSUE TITLE
One example of a much more elegant and concise code

### ISSUE THREAD

**Issue by tesla-cat** [neuromatch/NeuroAI_Course#436]
- For [Tutorial 2: Generalization in Neuroscience](https://neuroai.neuromatch.io/tutorials/W1D1_Generalization/student/W1D1_Tutorial2.html)

## My code below (**63 lines**) downloads, loads, plots the data

![Image](https://github.com/user-attachments/assets/f856dacb-5eda-4816-acae-e8ea05717e8f)

```py
import os

import matplotlib.pyplot as plt
import numpy as np
import requests
import scipy.io
import torch as tc


def download_data():
    files = [
        ("regularized_model_final.pth", "https://osf.io/kc7sb/download"),
        ("unregularized_model_final.pth", "https://osf.io/9vsy5/download"),
        ("condsForSimJ2moMuscles.mat", "https://osf.io/wak7e/download"),
        ("m1_reaching_data.mat", "https://osf.io/p2x4n/download"),
    ]
    os.mkdir("data")
    for name, url in files:
        with open(f"data/{name}", "wb+") as f:
            f.write(requests.get(url).content)


def tensor(x):
    return tc.tensor(np.array(x), dtype=tc.float32)


def normalize(x: tc.Tensor):
    a, b = x.min(), x.max()
    x = (x - a) / (b - a)
    return (x - x.mean()) / x.std()


def load_data(path):
    data = scipy.io.loadmat(path)
    d: np.ndarray = data["condsForSim"]

    def get(k):
        I, J = d.shape
        return tensor([[d[i, j][k] for j in range(J)] for i in range(I)])

    go, plan, mus = map(get, ("goEnvelope", "plan", "muscle"))
    plan = normalize(plan)
    inp = normalize(tc.cat([plan, go], dim=3))

    a, b, c = 3, slice(46, 296), [3, 4]
    return inp[:, a, b, :], mus[:, a, b, c]


def plot_data(inp: tc.Tensor, mus: tc.Tensor):
    cond = [0, 6, 9]
    R, C = 2, len(cond)
    plt.figure(figsize=(4 * C, 3 * R))
    for j, c in enumerate(cond):
        N = inp.shape[2]
        plt.subplot(R, C, j + 1)
        for i in range(N):
            a = 5 if i == N - 1 else 1
            plt.plot(a * inp[c, :, i])
        plt.subplot(R, C, j + 1 + C)
        for i in range(mus.shape[2]):
            plt.plot(mus[c, :, i])
    plt.tight_layout()
    plt.savefig("W1D1")


if __name__ == "__main__":
    # download_data()
    path = "data/condsForSimJ2moMuscles.mat"
    inp, mus = load_data(path)
    print(inp.shape, mus.shape)
    plot_data(inp, mus)

```

## The same involved **254** lines (with [black formatter](https://github.com/psf/black)) in the tutorial !

```py
import hashlib
import os

import matplotlib.pyplot as plt
import numpy as np
import requests
import scipy.io
import torch as tc


def retrieve_file(fname, url, expected_md5):
    # Check if the file already exists
    if not os.path.isfile(fname):
        try:
            # Attempt to download the file
            response = requests.get(url)
        except requests.ConnectionError:
            # Handle connection errors during the download
            print("!!! Failed to download data !!!")
        else:
            # No connection errors, proceed to check the response
            if response.status_code != requests.codes.ok:
                # Check if the HTTP response status code indicates a successful download
                print("!!! Failed to download data !!!")
            elif hashlib.md5(response.content).hexdigest() != expected_md5:
                # Verify the integrity of the downloaded file using MD5 checksum
                print("!!! Data download appears corrupted !!!")
            else:
                # If download is successful and data is not corrupted, save the file
                with open(fname, "wb") as fid:
                    fid.write(
                        response.content
                    )  # Write the downloaded content to a file


# List of files to be downloaded with their respective URLs and expected MD5 hashes
files = [
    (
        "regularized_model_final.pth",
        "https://osf.io/kc7sb/download",
        "9435a9c2ea75766144bf840b25bfb97e",
    ),
    (
        "unregularized_model_final.pth",
        "https://osf.io/9vsy5/download",
        "2e3dc9551b677206e2315788df354a91",
    ),
    (
        "condsForSimJ2moMuscles.mat",
        "https://osf.io/wak7e/download",
        "257d16c4d92759d615bf5cac75dd9a1f",
    ),
    (
        "m1_reaching_data.mat",
        "https://osf.io/p2x4n/download",
        "6fc65443b9632db47772dd2efaadeee0",
    ),
]

for fname, url, expected_md5 in files:
    retrieve_file(fname, url, expected_md5)


def normalize_and_standardize(tensor):
    """
    Normalize and standardize a given tensor.

    Args:
    - tensor: Tensor, the tensor to be normalized and standardized.

    Returns:
    - standardized_normalized_tensor: Tensor, the normalized and standardized tensor.
    """
    min_val, max_val = tensor.min(), tensor.max()
    tensor = (tensor - min_val) / (max_val - min_val)  # Normalize
    mean, std = tensor.mean(), tensor.std()
    return (tensor - mean) / std  # Standardize


def prepare_dataset(file_path, feature_idx=7, muscle_idx=1):
    """
    Load and preprocess data from a .mat file for RNN training.

    Args:
    - file_path: str, path to the .mat file containing the dataset.
    - feature_idx: int, index for individual features for plotting. Max 14.
    - muscle_idx: int, index for muscles for plotting. Max 1.

    Returns:
    - normalised_inputs: Tensor, normalized and concatenated Plan and Go Envelope tensors.
    - avg_output: Tensor, average muscle activity across conditions and delays.
    - timesteps: np.ndarray, array of time steps for plotting.
    """
    # Load the .mat file
    data = scipy.io.loadmat(file_path)

    # Extract condsForSim struct
    conds_for_sim = data["condsForSim"]

    # Initialize lists to store data for all conditions
    go_envelope_all, plan_all, muscle_all = [], [], []

    # Get the number of conditions (rows) and delay durations (columns)
    num_conditions, num_delays = conds_for_sim.shape

    times = conds_for_sim["timesREmove"][0][0] / 1000.0

    # Select the same time period as the PSTHs
    rg = slice(46, 296)

    for i in range(num_conditions):  # Loop through each condition
        go_envelope_condition, plan_condition, muscle_condition = [], [], []

        for j in range(num_delays):  # Loop through each delay duration
            condition = conds_for_sim[i, j]
            go_envelope, plan, muscle = (
                condition["goEnvelope"],
                condition["plan"],
                condition["muscle"],
            )
            selected_muscle_data = muscle[:, [3, 4]]  # Select only specific muscles
            go_envelope_condition.append(go_envelope[rg, :])
            plan_condition.append(plan[rg, :])
            muscle_condition.append(selected_muscle_data[rg, :])

        # Convert lists of arrays to tensors and append to all conditions
        go_envelope_all.append(
            torch.tensor(np.array(go_envelope_condition), dtype=torch.float32)
        )
        plan_all.append(torch.tensor(np.array(plan_condition), dtype=torch.float32))
        muscle_all.append(torch.tensor(np.array(muscle_condition), dtype=torch.float32))

    times = times[rg]

    # Stack tensors for all conditions
    go_envelope_tensor, plan_tensor, output = (
        torch.stack(go_envelope_all),
        torch.stack(plan_all),
        torch.stack(muscle_all),
    )

    # Cleanup to free memory
    del data, conds_for_sim, go_envelope_all, plan_all, muscle_all
    gc.collect()

    # Normalize and Standardize Plan Tensor
    plan_tensor = normalize_and_standardize(plan_tensor)

    # Normalise and concatenate Plan and Go Envelope Tensors
    normalised_inputs = normalize_and_standardize(
        torch.cat([plan_tensor, go_envelope_tensor], dim=3)
    )

    fixed_delay = 3
    inputs_no_delay = normalised_inputs[:, fixed_delay, ...]
    output_no_delay = output[:, fixed_delay, ...]
    return inputs_no_delay, normalised_inputs, output_no_delay, output, times


# Averaging across conditions and delays
reach_directions = [0, 6, 9]
for reach in reach_directions:
    # Plot inputs and outputs
    one_direction = normalised_inputs[reach, ...].clone()
    # Exaggerate the Go Cue for visualization purposes
    one_direction[:, -1] *= 5
    plot_inputs_over_time(
        times, one_direction, title=f"Inputs over Time, reach direction {reach}"
    )
    plot_muscles_over_time(
        times, outputs[reach, ...], title=f"Outputs over Time, reach direction {reach}"
    )


def plot_inputs_over_time(timesteps, avg_inputs, title="Inputs over Time"):
    """
    Plot the inputs over time.

    Inputs:
    - timesteps (list or array-like): A sequence of time steps at which the inputs were recorded.
      This acts as the x-axis in the plot, representing the progression of time.
    - avg_inputs (list or array-like): The average values of inputs corresponding to each time step.
      These values are plotted on the y-axis, showing the magnitude of inputs over time.
    - title (string): The title of the plot

    Returns:
    This function generates and displays a plot using Matplotlib.
    """

    with plt.xkcd():
        plt.figure(figsize=(8, 3))
        num_features = (
            avg_inputs.shape[1] if hasattr(avg_inputs, "shape") else len(avg_inputs[0])
        )

        for feature_idx in range(num_features):
            current_feature_values = (
                avg_inputs[:, feature_idx]
                if hasattr(avg_inputs, "shape")
                else [row[feature_idx] for row in avg_inputs]
            )
            label = (
                f"Feature {feature_idx + 1}"
                if feature_idx < num_features - 1
                else "Go Cue"
            )
            plt.plot(timesteps, current_feature_values, label=label)

        plt.title(title)
        plt.xlabel("Time (s)")
        plt.ylabel("Value (A.U.)")
        plt.subplots_adjust(right=0.7)
        plt.legend(loc="center left", bbox_to_anchor=(1, 0.5), fontsize=8)
        plt.tight_layout()
        plt.xlim(min(timesteps), max(timesteps))
        plt.show()


def plot_muscles_over_time(timesteps, avg_output, title="Muscles over Time"):
    """
    Plot the average outputs over time for two muscles to visualize changes in output values.
    The avg_output is expected to be a 250x2 array where each column corresponds to a different muscle.

    Inputs:
    - timesteps (list or array-like): A sequence of time steps at which the outputs were recorded.
      This acts as the x-axis in the plot, representing the progression of time.
    - avg_output (array-like, shape [250, 2]): The average values of outputs, with each column
      representing the output over time for each muscle.
    - title (string): The title of the plot

    Returns:
    This function generates and displays a plot using Matplotlib.
    """

    with plt.xkcd():

        plt.figure(figsize=(8, 3))  # Set the figure size
        plt.plot(timesteps, avg_output[:, 0], label="Muscle 1")  # Plot for muscle 1
        plt.plot(timesteps, avg_output[:, 1], label="Muscle 2")  # Plot for muscle 2
        plt.title(title)
        plt.xlabel("Time (s)")
        plt.ylabel("Value (A.U.)")

        # Adjust plot margins to provide space for the legend outside the plot
        plt.subplots_adjust(right=0.7)
        plt.legend(
            loc="center left", bbox_to_anchor=(1, 0.5), fontsize=8
        )  # Placing legend outside

        plt.tight_layout()
        plt.xlim(
            min(timesteps), max(timesteps)
        )  # Ensuring x-axis covers the range of timesteps
        plt.show()

```

**Comment by Alxmrphi** [comment]
Hi @tesla-cat, it's great you were able to create a more concise set of code and still get everything to run. However, being able to do (almost) the same amount in fewer lines isn't always the best solution, especially in the pedagogical world of being beginner-friendly and clear with code intentions. It's not enterprise-level code that would prioritise speed and fancy written code over something longer and more functional. 

Your code has no docstrings (which make up many lines in the official version we have) and it also has no comments. Docstrings and comments are very useful for students who want to study the exact consequence of each line of code. Additionally, there are defensive statements to prevent and warn against errors, which are not present in your code. Your code also doesn't document the figure in the way the original code does (style, axis labels, legend etc.)

While it's great to see you've been able to recreate this section yourself, I don't believe your proposal is sufficient to warrant replacement of what we have already, because it's less user-friendly, doesn't handle errors / exceptions, is missing important context (code comments and full docstrings) as well as not fully recreating the figure.

We would be doing the students a disservice to replace the current code with examples such as:
```python
    a, b, c = 3, slice(46, 296), [3, 4]
    return inp[:, a, b, :], mus[:, a, b, c]
```

where there is no explanation for what the axis dimensions are or where those values come from. 

Longer code with docstrings, commented lines and clear variable names is much more maintainable and friendly, especially within the environment where students are learning and where we want them to be able to follow along and understand code operations. Planning in good Exception raising and pre-empting errors is also a practice we want to convey to students. 

But we're happy you were able to re-code this yourself and get the same results :)
